src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs

```diff
@@ -82,56 +81,39 @@ public static object ConvertFromJson(string input, bool returnHashTable, out Err
                         MaxDepth = 1024
                     });
 
-                // JObject is a IDictionary
-                var dictionary = obj as JObject;
-                if (dictionary != null)
+                switch (obj)
                 {
-                    if (returnHashTable)
-                    {
-                        obj = PopulateHashTableFromJDictionary(dictionary, out error);
-                    }
-                    else
-                    {
-                        obj = PopulateFromJDictionary(dictionary, out error);
-                    }
-                }
-                else
-                {
-                    // JArray is a collection
-                    var list = obj as JArray;
-                    if (list != null)
-                    {
-                        if (returnHashTable)
-                        {
-                            obj = PopulateHashTableFromJArray(list, out error);
-                        }
-                        else
-                        {
-                            obj = PopulateFromJArray(list, out error);
-                        }
-                    }
+                    case JObject dictionary:
+                        // JObject is a IDictionary
+                        return returnHashTable
+                                   ? PopulateHashTableFromJDictionary(dictionary, out error)
+                                   : PopulateFromJDictionary(dictionary, new DuplicateMemberSet(dictionary.Count), out error);
+                    case JArray list:
+                        return returnHashTable
+                                   ? PopulateHashTableFromJArray(list, out error)
+                                   : PopulateFromJArray(list, out error);
+                    default: return obj;
                 }
             }
             catch (JsonException je)
             {
                 var msg = string.Format(CultureInfo.CurrentCulture, WebCmdletStrings.JsonDeserializationFailed, je.Message);
+
                 // the same as JavaScriptSerializer does
                 throw new ArgumentException(msg, je);
             }
-            return obj;
         }
 
         // This function is a clone of PopulateFromDictionary using JObject as an input.
-        private static PSObject PopulateFromJDictionary(JObject entries, out ErrorRecord error)
+        private static PSObject PopulateFromJDictionary(JObject entries, DuplicateMemberSet memberTracker, out ErrorRecord error)
         {
             error = null;
-            PSObject result = new PSObject();
+            var result = new PSObject(entries.Count);
             foreach (var entry in entries)
```

Comment:
Why?

Reply:
The type is clearly stated by the ctor. Why repeat it?

Our coding guidelines has 

```csharp
var example = obj as Example;
```

as an example. I don't see how this is any different. There can be no confusion as to what type result has. 

In general, I think it would be better to consider types as something the compiler uses to ensure we don't use something in an unintended way. 

Personally, I find all the type annotations reducing readability, since they get in the way of the essence.

Reply:
Again, as background, I really recommend Herb Sutters https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/

Reply:
Also, when looking at high quality, modern, C# code bases this is a very common, and in my eyes non-controversial.

https://github.com/dotnet/roslyn/blob/acd31862be412fcb4f7e0bc7a7db63ea8228a327/src/Compilers/Shared/DesktopBuildClient.cs#L39

https://github.com/dotnet/roslyn/blob/master/src/Compilers/Shared/BuildServerConnection.cs

Reply:
Do we both say about `foreach (var entry in entries) ` ?

Reply:
I don't think that the type of `entry` is of interest here. It is some entity that has a `Key` and a `Value`. Should that type change, nothing in the code would need to change, as long as it has a `Key` and a `Value`. 
That is captured nicely with the use of  `var`. 

I think we should only specify the type explicitly when it is important for some reason that the value is of a certain type. That way, it calls attention to it - something interesting is happening here since we explicitly specified the type. It may be that a variable is declared as a base type instead of the actual type returned by a function.

The C# coding conventions on [msdn](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/inside-a-program/coding-conventions) has actually captured this:
> Use implicit typing for local variables when the type of the variable is obvious from the right side of the assignment, *or when the precise type is not important.*

> Use implicit typing to determine the type of the loop variable in for and foreach loops. 
The semantics would also change.